### DISCUSSION TITLE
Feature Request: Support multiple regex / wildcards in `SELECT COLUMNS` and `SELECT * EXCLUDE`

### DISCUSSION THREAD

**Discussion by earlev4** [duckdb/duckdb#6374] [Ideas]
Hi DuckDB Community! **_Huge thanks and appreciation to the DuckDB creators and contributors. Thank you for all you do!!!_**

Depending on the situation, selecting needed columns can be easily achieved either using `SELECT COLUMNS` or `SELECT * EXCLUDE`. However, there are situations that require more versatility in using these clauses, especially when multiple conditions exists. 

Proposing to build on the "tremendous flexibility" that DuckDB already offers.

Using the attached CSV for BackBlaze hard drive data on 2022-10-01 as example:
[2022-10-01.zip](https://github.com/duckdb/duckdb/files/10777699/2022-10-01.zip)

Currently, `SELECT COLUMNS` allows for a regex pattern to select columns that match a pattern. Proposing to accept multiple regex patterns. 
Example query: 
```python
con.execute(
    """
    SELECT COLUMNS('smart_1_', 'smart_2_') 
    FROM read_csv_auto('2022-10-01.csv') LIMIT 5;
    """
).df()
```
This would select all columns that **match** either `smart_1_` or `smart_2_`.
Example output: 
```python
| smart_1_normalized | smart_1_raw | smart_2_normalized | smart_2_raw |
|-------------------:|------------:|-------------------:|------------:|
|                100 |           0 |                134 |         101 |
|                100 |           0 |                133 |         104 |
|                100 |           0 |                134 |         102 |
|                100 |           0 |                135 |          99 |
|                100 |           0 |                135 |          99 |
```

Currently, `EXCLUDE` allows column names to exclude from a query. Proposing to accept multiple regex patterns.
Query:
```python
    SELECT * EXCLUDE('model', '_normalized') 
    FROM read_csv_auto('2022-10-01.csv') LIMIT 5;
```
This would select all columns that do **not match** either `model` or `_normalized`.
Example output: 
```python
|       date |  serial_number | capacity_bytes | failure | smart_1_raw | ... | smart_255_raw |
|-----------:|---------------:|---------------:|--------:|------------:|----:|--------------:|
| 2022-10-01 | PL1331LAGDJ5GH |  4000787030016 |       0 |           0 | ... |           NaN |
| 2022-10-01 | PL2331LAGLW5UJ |  4000787030016 |       0 |           0 | ... |           NaN |
| 2022-10-01 | PL2331LAGM410J |  4000787030016 |       0 |           0 | ... |           NaN |
| 2022-10-01 | PL2331LAGM5NTJ |  4000787030016 |       0 |           0 | ... |           NaN |
| 2022-10-01 | PL2331LAGM5R4J |  4000787030016 |       0 |           0 | ... |           NaN |
```
This should align with the "Friendlier SQL" approach of: "save time repeatedly typing all columns, improve code readability, and retain flexibility". 

**_Looking forward to feedback. Thank you for your consideration. I sincerely appreciate it._**

**Comment by rsund** [comment]
Currently `EXCLUDE` seems not to accept regexes, so allowing such a feature would certainly be a very nice addition. However, changing the list of strings to list of regexes would change the behaviour in some cases, e.g. if there are columns model and model_1, then EXCLUDE ('model') using regexes would remove both in spite of the fact that with current behaviour without regexes model_1 will remain.

But as `COLUMNS` accepts regex, is there some particular case that it is not enough? You most likely noticed that your example query could be easily formulated as
```sql
SELECT COLUMNS('smart_[12]_') FROM read_csv_auto('2022-10-01.csv') LIMIT 5;
```
or 
```sql
SELECT COLUMNS('smart_1_|smart_2_') FROM read_csv_auto('2022-10-01.csv') LIMIT 5
```
where the latter one would essentially be the same as having "several regexes".

**Reply by earlev4** [reply]
Hi @rsund! Good points made! I appreciate the response and insightful feedback. 

I see your point about the behavior of the `EXCLUDE`. Using the 'model' example you provided, that would present challenges. If `EXCLUDE` was only to accept regex, could a boundary anchor be used where only 'model' column would be excluded?

```sql
SELECT * EXCLUDE('\\bmodel\\b') FROM read_csv_auto('2022-10-01.csv') LIMIT 5;
```

Regarding the `COLUMNS` accepting regex, those are great examples you provided of formulating the regex. I was thinking that combining regex patterns can be quickly difficult to read and understand, leading to more complexity. Breaking down to multiple smaller regex patterns (as an option, not required - still could use a more elaborate regex pattern, if wanted) lends to a "Friendlier SQL" where there might be improvements in code readability and offer greater flexibility.

Thanks so much for the feedback!

**Reply by earlev4** [reply]
Hi @rsund! After further consideration, changing the behavior of `EXCLUDE` seems problematic. There is already an expectation for `EXCLUDE` to accept columns and changing the behavior to accept regex patterns would break the current usage. Your feedback was very helpful in coming to this conclusion.

However, since the expectation of `COLUMNS` is to use regex patterns, could this clause be modified with a `NOT` to exclude columns matching a regex pattern. 

For example, the following would return all columns except those matching `smart_1_` or `smart_2_`.

```sql
SELECT NOT COLUMNS ('smart_1_|smart_2_') FROM read_csv_auto('2022-10-01.csv') LIMIT 5;
```

Or if smaller/simpler regex patterns were accepted:

```sql
SELECT NOT COLUMNS ('smart_1_', 'smart_2_') FROM read_csv_auto('2022-10-01.csv') LIMIT 5;
```

Thank you again for your feedback and the opportunity to brainstorm on this issue. I sincerely appreciate it.

**Reply by redviking1** [reply]
Might DuckDB resolve the issue with `EXCLUDE` by introducing a new function or keyword `EXCLUDE_REGEXP`?

**Reply by redviking1** [reply]
Another solution might be to introduce syntax `SELECT * EXCLUDE SIMILAR TO 'regular_expression'` and `SELECT COLUMNS SIMILAR TO 'regular_expression'` since `SIMILIR TO` pattern matching syntax and behaviour already exists.

**Reply by rsund** [reply]
Let's wait for the core team to comment on this idea.

I just liked the idea and wanted to support it and simultaneously ask questions that probably need to be solved in order to get this feature implemented. It seems that COLUMN -syntax needs also some other tuning (#6332), so hopefully this kind of feature could be also added at the same time.

In any case, having the possibility to use regexes for selecting and/or excluding certain columns (hopefully even so that both could be used at the same time) would be very nice addition. I'm not sure what would be the best syntax as there appears to be many possibilities (also including LIKE and NOT LIKE or something like COLUMNS('regex_include', EXCLUDE ('regex_exclude')). It certainly would be easiest to follow some other implementation, but I don't know is the COLUMNS() -syntax adopted from somewhere or is it just DuckDB's own syntactic sugar.

**Reply by earlev4** [reply]
Great discussion! I appreciate the interest and dialogue. Thanks for all the input @rsund and @redviking1! Greatly looking forward to the core team's comments on the idea! 😄

**Comment by Mytherin** [comment]
I have thought about this for a bit and I think the cleanest solution is to allow `COLUMNS` to support constant list arguments, for example:

```sql
SELECT COLUMNS(['a', 'b']) FROM tbl;
```

That way we can use list operations to apply essentially any filter available in the system. For example:

```sql
SELECT COLUMNS([x for x in * if x SIMILAR TO '...']) FROM tbl`
```

Or:

```sql
SELECT COLUMNS(LIST_FILTER(*, x -> x SIMILAR TO '...')) FROM tbl`
```

No additional syntax would have to be added to support a specific special case - and many more opportunities open up to apply all kinds of expressions/filters to column names.

**Reply by rsund** [reply]
Very elegant solution indeed!

**Reply by earlev4** [reply]
Very elegant and thoughtful solution, @Mytherin! I agree this approach offers many opportunities. **Thank you very much! I sincerely appreciate it.**

**Reply by Mytherin** [reply]
I implemented this in #6621, the syntax is somewhat improved:

```sql
SELECT COLUMNS(x -> x SIMILAR TO '...') FROM tbl`
```

**Reply by earlev4** [reply]
Thanks so much, @Mytherin!!! I greatly appreciate the implementation!!!
